flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/streaming/connectors/pulsar/FlinkPulsarProducer.java

```diff
@@ -0,0 +1,304 @@
+			LOG.warn("Flushing on checkpoint is enabled, but checkpointing is not enabled. Disabling flushing.");
+			flushOnCheckpoint = false;
+		}
+
+		if (PulsarProduceMode.AT_MOST_ONCE == produceMode) {
+			this.failureCallback = cause -> {
+				LOG.error("Error while sending record to Pulsar : " + cause.getMessage(), cause);
+				return null;
+			};
+		} else if (PulsarProduceMode.AT_LEAST_ONE == produceMode){
+			this.failureCallback = cause -> {
+				if (null == asyncException) {
+					if (cause instanceof Exception) {
+						asyncException = (Exception) cause;
+					} else {
+						asyncException = new Exception(cause);
+					}
+				}
+				return null;
+			};
+		} else {
+			throw new UnsupportedOperationException("Unsupported produce mode " + produceMode);
+		}
+	}
+
+	@Override
+	public void invoke(IN value, Context context) throws Exception {
+		checkErroneous();
+
+		byte[] serializedValue = schema.serialize(value);
+
+		MessageBuilder msgBuilder = MessageBuilder.create();
+		if (null != context.timestamp()) {
+			msgBuilder = msgBuilder.setEventTime(context.timestamp());
+		}
+		String msgKey = flinkPulsarKeyExtractor.getKey(value);
+		if (null != msgKey) {
+			msgBuilder = msgBuilder.setKey(msgKey);
+		}
+		Message message = msgBuilder
+			.setContent(serializedValue)
+			.build();
+
+		if (flushOnCheckpoint) {
+			synchronized (pendingRecordsLock) {
+				pendingRecords++;
+			}
+		}
+		producer.sendAsync(message)
+			.thenApply(successCallback)
+			.exceptionally(failureCallback);
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (producer != null) {
+			producer.close();
+		}
+
+		// make sure we propagate pending errors
+		checkErroneous();
+	}
+
+	// ------------------- Logic for handling checkpoint flushing -------------------------- //
+
+	private void acknowledgeMessage() {
+		if (flushOnCheckpoint) {
+			synchronized (pendingRecordsLock) {
+				pendingRecords--;
+				if (pendingRecords == 0) {
+					pendingRecordsLock.notifyAll();
+				}
+			}
+		}
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		// check for asynchronous errors and fail the checkpoint if necessary
+		checkErroneous();
+
+		if (flushOnCheckpoint) {
+			// wait until all the messages are acknowledged
+			synchronized (pendingRecordsLock) {
+				while (pendingRecords > 0) {
+					pendingRecordsLock.wait(100);
+				}
+			}
+
+			// if the flushed requests has errors, we should propagate it also and fail the checkpoint
+			checkErroneous();
+		}
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		// nothing to do
+	}
+
+	// ----------------------------------- Utilities --------------------------
+
+	protected void checkErroneous() throws Exception {
+		Exception e = asyncException;
+		if (e != null) {
+			// prevent double throwing
+			asyncException = null;
+			throw new Exception("Failed to send data to Kafka: " + e.getMessage(), e);
```

Comment:
Seems like typo is here.

Reply:
Yes, thanks for pointing out this mistake, it's correct in the newest commit.
ping @surryr